src/jit/lsrabuild.cpp

```diff
@@ -1311,542 +1471,79 @@ void LinearScan::buildRefPositionsForNode(GenTree* tree, BasicBlock* block, Lsra
-            // not-quite-correct setting of lastUse.  However, this is OK because
-            // 1) this is only for preferencing, which doesn't require strict correctness, and
-            //    for determing which largeVectors require having their upper-half saved & restored.
-            //    (Issue #17481 tracks the issue that this system results in excessive spills and
-            //    should be changed.)
-            // 2) the cases where these out-of-order uses occur should not overlap a kill (they are
-            //    only known to occur within a single expression).
-            if ((useNode->gtFlags & GTF_VAR_DEATH) != 0)
-            {
-                VarSetOps::RemoveElemD(compiler, currentLiveVars, srcInterval->getVarIndex(compiler));
-            }
-            refPosNode = useNode;
-        }
-        else
-        {
-            // For non-localVar uses we record nothing, as nothing needs to be written back to the tree.
-            refPosNode = nullptr;
-        }
-
-        RefPosition* pos = newRefPosition(srcInterval, currentLoc, RefTypeUse, refPosNode, candidates, 0);
-        if (delayRegFree)
-        {
-            pos->delayRegFree = true;
-        }
-
-        if (useNode->IsRegOptional())
-        {
-            pos->setAllocateIfProfitable(true);
-        }
-        consumed++;
-
-        // Create additional use RefPositions for multi-reg nodes.
-        for (int idx = 1; idx < locInfo.info.dstCount; idx++)
-        {
-            noway_assert(srcInterval->relatedInterval != nullptr);
-            srcInterval = srcInterval->relatedInterval;
-#ifdef _TARGET_ARM_
-            if (useNode->OperIsPutArgSplit() ||
-                (compiler->opts.compUseSoftFP && (useNode->OperIsPutArgReg() || useNode->OperGet() == GT_BITCAST)))
-            {
-                // get first candidate, set bits in useCandidates must be in sequential order.
-                candidates = genFindLowestReg(allCandidates);
-                allCandidates &= ~candidates;
-            }
-#endif // _TARGET_ARM_
-            RefPosition* pos = newRefPosition(srcInterval, currentLoc, RefTypeUse, refPosNode, candidates, idx);
-            consumed++;
-        }
-    }
-
-    assert(consumed == consume);
-    if (consume != 0)
-    {
-        listNodePool.ReturnNodes(useList);
-    }
-
-    buildInternalRegisterUsesForNode(tree, info, internalRefs, internalCount);
-
-    RegisterType registerType  = getDefType(tree);
-    regMaskTP    candidates    = info->getDstCandidates(this);
-    regMaskTP    useCandidates = info->getSrcCandidates(this);
+    int consume = BuildNode(tree);
 
 #ifdef DEBUG
-    if (VERBOSE && produce)
-    {
-        printf("Def candidates ");
-        dumpRegMask(candidates);
-        printf(", Use candidates ");
-        dumpRegMask(useCandidates);
-        printf("\n");
-    }
-#endif // DEBUG
+    int newDefListCount = defList.Count();
+    int produce         = newDefListCount - oldDefListCount;
```

Comment:
So the only use (and assert using) produce is for amd64? Should you define this below, then?

Reply:
This assert was pre-existing, though before there were other uses of `produce`. However, I'd like to strengthen this assert to apply more broadly, but it requires some cleanup of the various nodes that produce multiple results. I've added a note to issue #13183 that deals with multi-reg nodes to strengthen this assert as well.